File: internal/slots/math.go
package slots

import (
	"context"
	"fmt"
	"math"
	"strings"
	"zeno/pkg/engine"
	"zeno/pkg/utils/coerce"

	"github.com/expr-lang/expr"
	"github.com/shopspring/decimal"
)

func RegisterMathSlots(eng *engine.Engine) {

	// ==========================================
	// 1. SLOT: MATH.CALC (General Math - Float64)
	// ==========================================
	eng.Register("math.calc", func(ctx context.Context, node *engine.Node, scope *engine.Scope) error {
		expressionStr := coerce.ToString(node.Value)
		target := "calc_result"

		// Support shorthand & attributes
		for _, c := range node.Children {
			if c.Name == "as" {
				target = strings.TrimPrefix(coerce.ToString(c.Value), "$")
			}
			if c.Name == "val" || c.Name == "expr" {
				expressionStr = coerce.ToString(c.Value)
			}
		}

		if expressionStr == "" {
			return fmt.Errorf("math.calc: expression is required")
		}

		// 1. Siapkan Environment
		env := make(map[string]interface{})

		// Copy variabel dari scope & AUTO-CONVERT angka
		for k, v := range scope.ToMap() {
			// Cek apakah value ini string yang isinya angka? (misal: "5", "10.5")
			if str, ok := v.(string); ok {
				if f, err := coerce.ToFloat64(str); err == nil {
					env[k] = f // Simpan sebagai Float agar bisa dihitung
				} else {
					env[k] = str // Biarkan string jika bukan angka (misal: "budi")
				}
			} else {
				env[k] = v // Value lain (int, bool, object) biarkan apa adanya
			}
		}

		// [UPGRADE] Inject Fungsi Matematika Standar
		env["ceil"] = math.Ceil
		env["floor"] = math.Floor
		env["round"] = math.Round
		env["abs"] = math.Abs
		env["max"] = math.Max
		env["min"] = math.Min
		env["sqrt"] = math.Sqrt
		env["pow"] = math.Pow

		// 2. Pre-processing ($ -> kosong)
		cleanExpr := strings.ReplaceAll(expressionStr, "$", "")

		// 3. Compile & Run
		program, err := expr.Compile(cleanExpr, expr.Env(env))
		if err != nil {
			return fmt.Errorf("math.calc: syntax error '%s': %v", expressionStr, err)
		}

		output, err := expr.Run(program, env)
		if err != nil {
			return fmt.Errorf("math.calc: runtime error: %v", err)
		}

		scope.Set(target, output)
		return nil
	}, engine.SlotMeta{
		Description: "Melakukan perhitungan matematika menggunakan ekspresi string.",
		Example:     "math.calc: ceil($total / 10)\n  as: $pages",
		ValueType:   "string",
		Inputs: map[string]engine.InputMeta{
			"as":   {Description: "Variabel penyimpan hasil", Required: false, Type: "string"},
			"val":  {Description: "Ekspresi matematika (jika tidak via value utama)", Required: false, Type: "string"},
			"expr": {Description: "Alias untuk val", Required: false, Type: "string"},
		},
	})

	// ==========================================
	// 2. SLOT: MONEY.CALC (Financial Math - Decimal)
	// ==========================================
	eng.Register("money.calc", func(ctx context.Context, node *engine.Node, scope *engine.Scope) error {
		expressionStr := coerce.ToString(node.Value)
		target := "money_result"

		for _, c := range node.Children {
			if c.Name == "as" {
				target = strings.TrimPrefix(coerce.ToString(c.Value), "$")
			}
			if c.Name == "val" {
				expressionStr = coerce.ToString(c.Value)
			}
		}

		// 1. Siapkan Environment Decimal
		env := make(map[string]interface{})

		for k, v := range scope.ToMap() {
			// Only convert if it looks like a number
			s := coerce.ToString(v)
			if s != "" && (s[0] == '-' || (s[0] >= '0' && s[0] <= '9')) {
				if d, err := decimal.NewFromString(s); err == nil {
					env[k] = d
					continue
				}
			}
			env[k] = v // Keep original for non-numeric
		}

		// Inject Decimal Functions for Operator Overloading
		env["Add"] = func(a, b decimal.Decimal) decimal.Decimal { return a.Add(b) }
		env["Sub"] = func(a, b decimal.Decimal) decimal.Decimal { return a.Sub(b) }
		env["Mul"] = func(a, b decimal.Decimal) decimal.Decimal { return a.Mul(b) }
		env["Div"] = func(a, b decimal.Decimal) decimal.Decimal { return a.Div(b) }

		// [UPGRADE] Inject Decimal Constructor for literals
		env["decimal"] = func(v interface{}) decimal.Decimal {
			d, _ := decimal.NewFromString(coerce.ToString(v))
			return d
		}

		// 2. Pre-processing
		cleanExpr := strings.ReplaceAll(expressionStr, "$", "")

		// 3. Konfigurasi Expr (Operator Overloading)
		options := []expr.Option{
			expr.Env(env),
			expr.Operator("+", "Add"),
			expr.Operator("-", "Sub"),
			expr.Operator("*", "Mul"),
			expr.Operator("/", "Div"),
		}

		program, err := expr.Compile(cleanExpr, options...)
		if err != nil {
			return fmt.Errorf("money.calc: syntax error '%s': %v", expressionStr, err)
		}

		output, err := expr.Run(program, env)
		if err != nil {
			return fmt.Errorf("money.calc: runtime error: %v", err)
		}

		// Return sebagai String
		if d, ok := output.(decimal.Decimal); ok {
			scope.Set(target, d.String())
		} else {
			scope.Set(target, coerce.ToString(output))
		}

		return nil
	}, engine.SlotMeta{
		Description: "Melakukan perhitungan keuangan menggunakan Decimal untuk presisi tinggi.",
		Example:     "money.calc: ($harga * $qty) - $diskon\n  as: $total",
		ValueType:   "decimal",
		Inputs: map[string]engine.InputMeta{
			"as":  {Description: "Variabel penyimpan hasil", Required: false, Type: "string"},
			"val": {Description: "Ekspresi keuangan", Required: false, Type: "decimal"},
		},
	})
}

File: zeno-rs/crates/zeno-std/src/math.rs
use zenocore::{Diagnostic, Engine, InputMeta, SlotMeta, Value};
use zenocore::slots::resolve_node_value;
use evalexpr::{ContextWithMutableFunctions, ContextWithMutableVariables};
use std::collections::HashMap;
use std::sync::Arc;

pub fn register(engine: &mut Engine) {
    // ==========================================
    // MATH.CALC
    // ==========================================
    engine.register(
        "math.calc",
        Arc::new(|engine, _ctx, node, scope| {
            let mut expression_str = String::new();
            let mut target = "calc_result".to_string();

            if node.value.is_some() {
                expression_str = resolve_node_value(engine, node, scope).to_string_coerce();
            }

            for c in &node.children {
                if c.name == "val" || c.name == "expr" {
                    expression_str = engine.resolve_shorthand_value(c, scope).to_string_coerce();
                }
                if c.name == "as" {
                    if let Some(ref cv) = c.value {
                        target = cv.trim_start_matches('$').to_string();
                    }
                }
            }

            if expression_str.is_empty() {
                return Err(Diagnostic {
                    r#type: "error".to_string(),
                    message: "math.calc: expression is required".to_string(),
                    filename: node.filename.clone(),
                    line: node.line,
                    col: node.col,
                    slot: Some("math.calc".to_string()),
                });
            }

            let clean_expr = expression_str.replace('$', "");
            let mut eval_context = evalexpr::HashMapContext::new();

            for (k, v) in scope.to_map() {
                let eval_val = match v {
                    Value::Nil => evalexpr::Value::Empty,
                    Value::Bool(b) => evalexpr::Value::Boolean(b),
                    Value::Int(i) => evalexpr::Value::Float(i as f64),
                    Value::Float(f) => evalexpr::Value::Float(f),
                    Value::String(s) => {
                        if let Ok(i) = s.parse::<i64>() {
                            evalexpr::Value::Float(i as f64)
                        } else if let Ok(f) = s.parse::<f64>() {
                            evalexpr::Value::Float(f)
                        } else {
                            evalexpr::Value::String(s)
                        }
                    }
                    _ => evalexpr::Value::Empty,
                };
                let _ = eval_context.set_value(k.into(), eval_val);
            }

            // Inject standard math functions
            let _ = eval_context.set_function("ceil".into(), evalexpr::Function::new(|v| {
                let f = match v {
                    evalexpr::Value::Float(f) => *f,
                    evalexpr::Value::Int(i) => *i as f64,
                    _ => return Err(evalexpr::EvalexprError::expected_number(v.clone())),
                };
                Ok(evalexpr::Value::Float(f.ceil()))
            }));
            let _ = eval_context.set_function("floor".into(), evalexpr::Function::new(|v| {
                let f = match v {
                    evalexpr::Value::Float(f) => *f,
                    evalexpr::Value::Int(i) => *i as f64,
                    _ => return Err(evalexpr::EvalexprError::expected_number(v.clone())),
                };
                Ok(evalexpr::Value::Float(f.floor()))
            }));
            let _ = eval_context.set_function("round".into(), evalexpr::Function::new(|v| {
                let f = match v {
                    evalexpr::Value::Float(f) => *f,
                    evalexpr::Value::Int(i) => *i as f64,
                    _ => return Err(evalexpr::EvalexprError::expected_number(v.clone())),
                };
                Ok(evalexpr::Value::Float(f.round()))
            }));
            let _ = eval_context.set_function("abs".into(), evalexpr::Function::new(|v| {
                match v {
                    evalexpr::Value::Float(f) => Ok(evalexpr::Value::Float((*f).abs())),
                    evalexpr::Value::Int(i) => Ok(evalexpr::Value::Int((*i).abs())),
                    _ => Err(evalexpr::EvalexprError::expected_number(v.clone())),
                }
            }));
            let _ = eval_context.set_function("sqrt".into(), evalexpr::Function::new(|v| {
                let f = match v {
                    evalexpr::Value::Float(f) => *f,
                    evalexpr::Value::Int(i) => *i as f64,
                    _ => return Err(evalexpr::EvalexprError::expected_number(v.clone())),
                };
                Ok(evalexpr::Value::Float(f.sqrt()))
            }));
            let _ = eval_context.set_function("pow".into(), evalexpr::Function::new(|v| {
                let args = match v {
                    evalexpr::Value::Tuple(vec) => vec,
                    _ => return Err(evalexpr::EvalexprError::expected_tuple(v.clone())),
                };
                if args.len() != 2 {
                    return Err(evalexpr::EvalexprError::expected_tuple(v.clone()));
                }
                let base = match &args[0] {
                    evalexpr::Value::Float(f) => *f,
                    evalexpr::Value::Int(i) => *i as f64,
                    _ => return Err(evalexpr::EvalexprError::expected_number(args[0].clone())),
                };
                let exp = match &args[1] {
                    evalexpr::Value::Float(f) => *f,
                    evalexpr::Value::Int(i) => *i as f64,
                    _ => return Err(evalexpr::EvalexprError::expected_number(args[1].clone())),
                };
                Ok(evalexpr::Value::Float(base.powf(exp)))
            }));
            let _ = eval_context.set_function("max".into(), evalexpr::Function::new(|v| {
                let args = match v {
                    evalexpr::Value::Tuple(vec) => vec,
                    _ => return Err(evalexpr::EvalexprError::expected_tuple(v.clone())),
                };
                if args.is_empty() {
                    return Err(evalexpr::EvalexprError::expected_tuple(v.clone()));
                }
                let mut max_val = f64::MIN;
                for arg in args {
                    let f = match arg {
                        evalexpr::Value::Float(f) => *f,
                        evalexpr::Value::Int(i) => *i as f64,
                        _ => return Err(evalexpr::EvalexprError::expected_number(arg.clone())),
                    };
                    if f > max_val {
                        max_val = f;
                    }
                }
                Ok(evalexpr::Value::Float(max_val))
            }));
            let _ = eval_context.set_function("min".into(), evalexpr::Function::new(|v| {
                let args = match v {
                    evalexpr::Value::Tuple(vec) => vec,
                    _ => return Err(evalexpr::EvalexprError::expected_tuple(v.clone())),
                };
                if args.is_empty() {
                    return Err(evalexpr::EvalexprError::expected_tuple(v.clone()));
                }
                let mut min_val = f64::MAX;
                for arg in args {
                    let f = match arg {
                        evalexpr::Value::Float(f) => *f,
                        evalexpr::Value::Int(i) => *i as f64,
                        _ => return Err(evalexpr::EvalexprError::expected_number(arg.clone())),
                    };
                    if f < min_val {
                        min_val = f;
                    }
                }
                Ok(evalexpr::Value::Float(min_val))
            }));

            let eval_result = evalexpr::eval_with_context(&clean_expr, &eval_context).map_err(|e| Diagnostic {
                r#type: "error".to_string(),
                message: format!("math.calc: error evaluating '{}': {}", expression_str, e),
                filename: node.filename.clone(),
                line: node.line,
                col: node.col,
                slot: Some("math.calc".to_string()),
            })?;

            let final_val = match eval_result {
                evalexpr::Value::Int(i) => Value::Int(i),
                evalexpr::Value::Float(f) => Value::Float(f),
                evalexpr::Value::Boolean(b) => Value::Bool(b),
                evalexpr::Value::String(s) => Value::String(s),
                _ => Value::Nil,
            };

            scope.set(&target, final_val);
            Ok(())
        }),
        SlotMeta {
            description: "Evaluate a mathematical expression string.".to_string(),
            example: "math.calc: ceil($total / 10)\n  as: $pages".to_string(),
            inputs: {
                let mut m = HashMap::new();
                m.insert("as".to_string(), InputMeta { description: "Target variable".to_string(), required: false, r#type: "any".to_string() });
                m.insert("val".to_string(), InputMeta { description: "Math expression".to_string(), required: false, r#type: "string".to_string() });
                m.insert("expr".to_string(), InputMeta { description: "Math expression".to_string(), required: false, r#type: "string".to_string() });
                m
            },
            required_blocks: Vec::new(),
            value_type: String::new(),
        },
    );

    // ==========================================
    // MONEY.CALC
    // ==========================================
    engine.register(
        "money.calc",
        Arc::new(|engine, _ctx, node, scope| {
            let mut expression_str = String::new();
            let mut target = "money_result".to_string();

            if node.value.is_some() {
                expression_str = resolve_node_value(engine, node, scope).to_string_coerce();
            }

            for c in &node.children {
                if c.name == "val" {
                    expression_str = engine.resolve_shorthand_value(c, scope).to_string_coerce();
                }
                if c.name == "as" {
                    if let Some(ref cv) = c.value {
                        target = cv.trim_start_matches('$').to_string();
                    }
                }
            }

            if expression_str.is_empty() {
                return Err(Diagnostic {
                    r#type: "error".to_string(),
                    message: "money.calc: expression is required".to_string(),
                    filename: node.filename.clone(),
                    line: node.line,
                    col: node.col,
                    slot: Some("money.calc".to_string()),
                });
            }

            let clean_expr = expression_str.replace('$', "");
            let mut eval_context = evalexpr::HashMapContext::new();

            for (k, v) in scope.to_map() {
                let eval_val = match v {
                    Value::Nil => evalexpr::Value::Empty,
                    Value::Bool(b) => evalexpr::Value::Boolean(b),
                    Value::Int(i) => evalexpr::Value::Int(i),
                    Value::Float(f) => evalexpr::Value::Float(f),
                    Value::String(s) => {
                        if let Ok(i) = s.parse::<i64>() {
                            evalexpr::Value::Int(i)
                        } else if let Ok(f) = s.parse::<f64>() {
                            evalexpr::Value::Float(f)
                        } else {
                            evalexpr::Value::String(s)
                        }
                    }
                    _ => evalexpr::Value::Empty,
                };
                let _ = eval_context.set_value(k.into(), eval_val);
            }

            let eval_result = evalexpr::eval_with_context(&clean_expr, &eval_context).map_err(|e| Diagnostic {
                r#type: "error".to_string(),
                message: format!("money.calc: error evaluating '{}': {}", expression_str, e),
                filename: node.filename.clone(),
                line: node.line,
                col: node.col,
                slot: Some("money.calc".to_string()),
            })?;

            let decimal_str = match eval_result {
                evalexpr::Value::Int(i) => {
                    use rust_decimal::prelude::FromPrimitive;
                    if let Some(d) = rust_decimal::Decimal::from_i64(i) {
                        d.to_string()
                    } else {
                        i.to_string()
                    }
                }
                evalexpr::Value::Float(f) => {
                    use rust_decimal::prelude::FromPrimitive;
                    if let Some(d) = rust_decimal::Decimal::from_f64(f) {
                        d.normalize().to_string()
                    } else {
                        f.to_string()
                    }
                }
                evalexpr::Value::String(s) => s,
                _ => "0".to_string(),
            };

            scope.set(&target, Value::String(decimal_str));
            Ok(())
        }),
        SlotMeta {
            description: "Financial math calculations using high precision decimals.".to_string(),
            example: "money.calc: ($harga * $qty) - $diskon\n  as: $total".to_string(),
            inputs: {
                let mut m = HashMap::new();
                m.insert("as".to_string(), InputMeta { description: "Target variable".to_string(), required: false, r#type: "any".to_string() });
                m.insert("val".to_string(), InputMeta { description: "Financial expression".to_string(), required: false, r#type: "string".to_string() });
                m
            },
            required_blocks: Vec::new(),
            value_type: String::new(),
        },
    );
}
